### PR TITLE
Support metrics in filtering expressions.

### PR DESCRIPTION
**Pull request by mstemm**
While working on the example programs for AWS/Cloudwatch, I found that
the Cloudwatch reads were pretty slow, as they had to fetch all
metrics, even if the read was really only interested in a single
metric like CPUUtilization.

To address this, add support for metrics in filtering expessions in
addition to items. This required a rework of FilterCloudWatchCompiler
to move beyond just a simple list of product= or item= filters. The
filtering expression now supports using AND to combine a
product=<product>, item=<item>, and/or metric=<metric> selector
together into a single condition. You can still use a concise format
of item=<product>:<item> or metric=<product>:<metric>, which is
essentially shorthand for product=<product> AND item=<item>. OR is
used to build up a list of conditions.

Within CloudWatchMon, it's given a list of conditions instead of a
list of products, as a product may appear in multiple conditions. If a
condition has a list of metrics, it is used to filter down the result
from listMetrics to only the desired metrics before calling
getMetricStatistics.

Add a bunch of unit tests that test all the new negative and positive
cases this introduces. While doing this, I found that some of the
negative tests were not actually working. In cases where the incorrect
filtering expression falsely compiled successfully, it wouldn't throw
an error at all and as a result wouldn't get to the expects() in the
catch handler. Fix that by throwing an error if compile doesn't.

@VladVega @davidvgalbraith

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -140,9 +134,40 @@ A product filter has the format `product="<aws product>"`, where `<aws product>`
...
+
+A metric filter has the format `metric="<metric>"`, specifying a specific metric (e.g. "CPUUtilization" for EC2, "VolumeReadBytes" for EBS). If any metric field is specified, only those CloudWatch metrics are returned.
+
+To combine products and items, use AND (e.g. product="EC2" and item="i-cc696a17"). You can also specify items and metrics using a concise format with the product included, using the form `item="<aws product>:<item name>"` or `metric="<aws product>:<metric name>"`.
+
+Other boolean logic such as NOT is not supported.
+
+Here are some example filter expressions:
+
+```Javascript
```
Comment (VladVega) [inline README.md]: Consider changing to `Juttle`.

File: lib/filter-cloudwatch-compiler.js [hunk 2]
```diff
@@ -68,80 +78,187 @@ var FilterCloudWatchCompiler = ASTVisitor.extend({
...
+                    this.throwUnsupportedFilter(node.location, 'AND between products');
+                }
+
+                if ((left.item.length > 0 && right.item.length > 0)) {
+                    this.throwUnsupportedFilter(node.location, 'AND between items');
+                }
+
+                if ((left.metric.length > 0 && right.metric.length > 0)) {
+                    this.throwUnsupportedFilter(node.location, 'AND between metrics');
+                }
```
Comment (VladVega) [inline lib/filter-cloudwatch-compiler.js]: Consider this instead of the above if statements:
```
_.each(['product', 'item', 'metric'], (type) => {
  if ((left[type].length > 0 && right[type].length > 0)) {
    this.throwUnsupportedFilter(node.location, `AND between ${type}s`);
  }
})
```

File: lib/filter-cloudwatch-compiler.js [hunk 3]
```diff
@@ -68,80 +78,187 @@ var FilterCloudWatchCompiler = ASTVisitor.extend({
...
                 right = this.visit(node.right);
 
-                // Left *must* be either 'item' or 'product'. When
-                // left is 'product', right must be one of the
-                // supported AWS products.
-                if (left !== 'item' && left !== 'product') {
+                // Left *must* be either 'item', 'product', or
+                // 'metric'. When left is 'product', right must be one
+                // of the supported AWS products.
+                if (left !== 'item' && left !== 'product' && left !== 'metric') {
```
Comment (VladVega) [inline lib/filter-cloudwatch-compiler.js]: Seems like a ```const CONDITION_KEYS = ['product', 'item', 'metric']``` at the top of the file might be useful.

Then `CONDITION_KEYS.indexOf(left) >= 0`

**Comment by VladVega** [comment]
Nice job, good testing, +1 from me.

File: lib/filter-cloudwatch-compiler.js (hunk 2 above)
Comment (mstemm) [inline lib/filter-cloudwatch-compiler.js]: I'll do that for item/metric, but product isn't an array so I think I'll keep as is.
